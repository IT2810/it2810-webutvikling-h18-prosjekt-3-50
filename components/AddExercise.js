import React, { Component } from 'react'
import { Container, Header, Content, Form, Item, Input, Button, Label, Card, CardItem, Row} from 'native-base';
import { View, Text, StyleSheet } from 'react-native'

class AddExercise extends Component {
  static navigationOptions = {
    title: 'AddExercise'
  }
  
  render() {

    return (
      <Container>
        <Content>
          <Form>
            <Item floatingLabel>
              <Label> Name </Label>
              <Input/>
            </Item>

            <Text>
              Sets
            </Text>
            
            <Card style={{flex: 0, flexWrap: 'wrap'}}>
              <CardItem>
                <Item floatingLabel style={{width: 50}}>
                  <Label> sets </Label>
                  <Input />
                </Item>

                <Text> x </Text>
                <Item floatingLabel style={{width: 50}}>
                  <Label> reps </Label>
                  <Input />
                </Item>

                
                <Item floatingLabel style={{width: 50}}>
                  <Label> kg </Label>
                  <Input />
                </Item>
              </CardItem>
            </Card>

            <Container style={styles.row}>
              <Button
                primary
              >
                <Text> Add exercise </Text>
              </Button>
              <Button
                light
              >
                <Text> Cancel </Text>
              </Button>
            </Container>
          </Form>
        </Content>
      </Container>
    )
  }
}

export default AddExercise

const styles = StyleSheet.create ({

  row: {
    display: 'flex',
    flexDirection: 'row' 
  }


})