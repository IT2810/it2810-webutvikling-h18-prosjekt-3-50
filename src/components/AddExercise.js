import React, { Component } from 'react'
import { View, Text, Container, Header, Content, Item, Input, Button, Label, Toast, Row, Fab, Icon, Form} from 'native-base';
import { StyleSheet, KeyboardAvoidingView } from 'react-native'

import { connect } from 'react-redux'

import { addExercise } from '../../actions/index'

export class AddExercise extends Component {
  static navigationOptions = {
    title: 'AddExercise'
  }

  constructor(props, context) {
    super(props, context)
    this.state = {
      name: null,
      reps: null,
      sets: null,
      kg: null
    }

    this.addExercise = this.addExercise.bind(this)
    this.validateExercise = this.validateExercise.bind(this)
    this.showToast = this.showToast.bind(this)
  }

  addExercise() {
    if (this.validateExercise()) {
      this.showToast('Added exercise', 'success')

      this.props.addExercise({
        name: this.state.name,
        reps: this.state.reps,
        sets: this.state.sets,
        kg: this.state.kg
      })
      this.props.navigation.navigate('CreateSession')
    } else {
      console.log("Exercise is not valid")
    }
  }

  validateExercise() {
    if (this.state.name != null) {
      if (this.state.reps == null || this.state.sets == null) {
        this.showToast('Sets and/or reps must be set', 'danger')
        return false
      } else {
        return true
      }
    } else {
      this.showToast('You have to give the exercise a name', 'danger')
      return false
    }
  }

  showToast(text, type) {
    Toast.show({
      type: type,
      duration: 3000,
      text: text,
      position: "top"
    })
  }

  render() {

    return (
      <KeyboardAvoidingView behavior="padding" style={{ flex: 1 }}>
        <Container testID={"container"}>
          <Content padder testID={"content"}>
            <Form>
              <View style={styles.inputView} testID={"view"}>
                <Text style={styles.inputTitle}>Description</Text>
                <Item>
                  <Input
                    onChangeText={(name) => {this.setState({name: name})}}
                    testID={"nameInput"}
                    value={this.state.name}
                    placeholder="Enter description here"
                  />
                </Item>
              </View>

              <View style={styles.inputView}>
                <Text style={styles.inputTitle}>Sets</Text>
                <Item>
                  <Input
                    onChangeText={(sets) => this.setState({sets})}
                    testID={"setsInput"}
                    value={this.state.sets}
                    placeholder="0"
                    keyboardType="numeric"
                  />
                </Item>
              </View>

              <View style={styles.inputView}>
                <Text style={styles.inputTitle}>Reps</Text>
                <Item>
                  <Input
                    onChangeText={(reps) => this.setState({reps})}
                    testID={"repsInput"}
                    value={this.state.reps}
                    placeholder="0"
                    keyboardType="numeric"
                  />
                </Item>
              </View>

              <View style={styles.inputView}>
                <Text style={styles.inputTitle}>Weight in kg (optional)</Text>
                <Item>
                  <Input
                    onChangeText={(kg) => this.setState({kg})}
                    testID={"kgInput"}
                    value={this.state.kg}
                    placeholder="0"
                    keyboardType="numeric"
                  />
                </Item>
              </View>
            </Form>
          </Content>
          <Fab
            onPress={this.addExercise}
            testID={'addExerciseButton'}
            style={{ backgroundColor: '#199E59' }}
            position="bottomRight"
          >
            <Icon ios="ios-checkmark" android="md-checkmark" style={{fontSize: 50, lineHeight: 50}} />

          </Fab>
        </Container>
      </KeyboardAvoidingView>
    )
  }
}

const mapDispatchToProps = dispatch => ({
  addExercise: exercise => dispatch(addExercise(exercise))
})

export default connect(null, mapDispatchToProps)(AddExercise)

const styles = StyleSheet.create ({
  inputView: {
    marginTop: '8%',
    paddingLeft: '4%',
    paddingRight: '4%'
  },
  inputTitle: {
    fontSize: 20,
    fontWeight: "bold"
  }
})
