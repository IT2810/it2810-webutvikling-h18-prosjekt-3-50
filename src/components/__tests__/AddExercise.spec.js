/**
 * @jest-environment jsdom
 */

import React from 'react'
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json';
import AddExercise from '../AddExercise'
import 'native-base'
import { findByID } from './testUtils.js'


import renderer from 'react-test-renderer';

describe('AddExercise', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallow(<AddExercise />)
  })

  it('renders correctly', () => {
    const wrapper = shallow(<AddExercise />).dive()

    expect(toJson(wrapper)).toMatchSnapshot()
  })
  

  describe('addExercise', () => {
    /*it('calls addExercise when button is pressed', () => {
      let addExerciseMock = jest.fn()
      let wrapper = shallow(<AddExercise />)

      //AddExercise.prototype.addExercise = addExerciseMock
      console.log("Before adding mocks")
      //wrapper.instance().addExercise = addExerciseMock //jest.fn(() => console.log("Mock function"))
      wrapper.instance().showToast = jest.fn()
      console.log(wrapper.instance().addExercise)
      console.log(addExerciseMock)


      wrapper.update()
      //let spy = jest.spyOn(wrapper.instance(), 'addExercise').mockImplementation(() => { console.log('second mock')})

      //console.log(wrapper.instance().addExercise)

      //wrapper.update()
      let spy = jest.spyOn(wrapper.instance(), 'addExercise')

      console.log(wrapper.instance().addExercise)
      console.log(spy)

      let button = findByID(wrapper, 'addExerciseButton')
      button.props().onPress()
      console.log("After press")
      expect(spy).toHaveBeenCalled()
      expect(wrapper.instance().addExercise).toHaveBeenCalled()
      expect(addExerciseMock).toHaveBeenCalled()
    })
*/
    it('navigates to SessionForm if validateExercise is true', () => {
      const navigateMock = jest.fn()
      const navigation = {navigate: navigateMock}
      const wrapper = shallow(<AddExercise navigation={navigation} />)

      let validateExerciseMock = jest.fn()
      validateExerciseMock.mockReturnValue(true)


      wrapper.instance().validateExercise = validateExerciseMock
      wrapper.update()

      wrapper.instance().addExercise()

      expect(navigateMock.mock.calls.length).toBe(1)
    })

    it('does not navigates to SessionForm if validateExercise is false', () => {
      const navigateMock = jest.fn()
      const navigation = {navigate: navigateMock}
      const wrapper = shallow(<AddExercise navigation={navigation} />)

      let validateExerciseMock = jest.fn()
      validateExerciseMock.mockReturnValue(false)


      wrapper.instance().validateExercise = validateExerciseMock
      wrapper.update()

      wrapper.instance().addExercise()

      expect(navigateMock.mock.calls.length).toBe(0)
    })
  })

  describe('state updated when giving input', () => {
    it('name of exercise', () => {
      let input = 'Squat'      
      let nameInput = findByID(wrapper, 'nameInput')
      nameInput.props().onChangeText(input)

      expect(wrapper.state('name')).toEqual(input)
    })

    it('sets of exercise', () => {
      let input = '4'      
      let sets = findByID(wrapper, 'setsInput')
      sets.props().onChangeText(input)

      expect(wrapper.state('sets')).toEqual(input)
    })

    it('reps of exercise', () => {
      let input = '12'      
      let repsInput = findByID(wrapper, 'repsInput')
      repsInput.props().onChangeText(input)

      expect(wrapper.state('reps')).toEqual(input)
    })

    it('kg of exercise', () => {
      let input = '60'      
      let kgInput = findByID(wrapper, 'kgInput')
      kgInput.props().onChangeText(input)

      expect(wrapper.state('kg')).toEqual(input)
    })

  })

  describe('validate', () => {

    it('exercise is valid if both name, sets and reps are given', () => {
      wrapper.instance().showToast = jest.fn()
      wrapper.state().name = 'TestName'
      wrapper.state().reps = '10'
      wrapper.state().sets = '4'

      expect(wrapper.instance().validateExercise()).toBeTruthy()
    })

    it('exercise is not valid if either name, sets and/or reps are missing', () => {
      wrapper.instance().showToast = jest.fn()

      wrapper.state().name = 'TestName'
      expect(wrapper.instance().validateExercise()).toBeFalsy()

      wrapper.state().reps = '10'
      expect(wrapper.instance().validateExercise()).toBeFalsy()

      wrapper.state().reps = null
      wrapper.state().sets = '4'
      expect(wrapper.instance().validateExercise()).toBeFalsy()
    })

  })
})
