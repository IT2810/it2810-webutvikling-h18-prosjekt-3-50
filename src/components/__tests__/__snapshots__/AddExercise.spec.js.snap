// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddExercise renders correctly 1`] = `
<Container
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
      "height": 1334,
    }
  }
  testID="container"
>
  <Styled(Content)
    padder={true}
    testID="content"
  >
    <Styled(ViewNB)
      padder={true}
      testID="view"
    >
      <Styled(Item)
        floatingLabel={true}
      >
        <Styled(Label)>
           Name 
        </Styled(Label)>
        <Styled(Input)
          onChangeText={[Function]}
          testID="nameInput"
          value={null}
        />
      </Styled(Item)>
    </Styled(ViewNB)>
    <Styled(ViewNB)>
      <Styled(Text)>
        Sets (weight is optional)
      </Styled(Text)>
      <RowNB
        style={
          Object {
            "flex": 0,
            "flexWrap": "wrap",
            "paddingBottom": 16,
          }
        }
      >
        <Styled(Item)
          floatingLabel={true}
          style={
            Object {
              "width": 50,
            }
          }
        >
          <Styled(Label)>
             sets 
          </Styled(Label)>
          <Styled(Input)
            onChangeText={[Function]}
            testID="setsInput"
            value={null}
          />
        </Styled(Item)>
        <Styled(Text)
          style={
            Object {
              "color": "grey",
              "fontSize": 16,
              "textAlignVertical": "bottom",
            }
          }
        >
           x 
        </Styled(Text)>
        <Styled(Item)
          floatingLabel={true}
          style={
            Object {
              "width": 50,
            }
          }
        >
          <Styled(Label)>
             reps 
          </Styled(Label)>
          <Styled(Input)
            onChangeText={[Function]}
            testID="repsInput"
            value={null}
          />
        </Styled(Item)>
        <Styled(Text)
          style={
            Object {
              "color": "grey",
              "fontSize": 16,
              "textAlignVertical": "bottom",
            }
          }
        >
           x 
        </Styled(Text)>
        <Styled(Item)
          floatingLabel={true}
          style={
            Object {
              "width": 50,
            }
          }
        >
          <Styled(Label)>
             kg 
          </Styled(Label)>
          <Styled(Input)
            onChangeText={[Function]}
            testID="kgInput"
            value={null}
          />
        </Styled(Item)>
      </RowNB>
      <Styled(Button)
        block={true}
        flat={true}
        onPress={[Function]}
        primary={true}
        testID="addExerciseButton"
      >
        <Styled(Text)>
           Add exercise 
        </Styled(Text)>
      </Styled(Button)>
    </Styled(ViewNB)>
  </Styled(Content)>
</Container>
`;
